### PR TITLE
ToEvent returns a pointer. Move Sender/Receiver out of Binding.

### PR DESCRIPTION
**Pull request by n3wscott**
Signed-off-by: Scott Nichols <snichols@vmware.com>

### CODE REVIEW

File: pkg/binding/to_event.go [hunk 1]
```diff
@@ -19,40 +19,41 @@ var ErrCannotConvertToEvent = errors.New("cannot convert message to event")
 // This function returns the Event generated from the Message and the original encoding of the message or
 // an error that points the conversion error.
 // transformers can be nil and this function guarantees that they are invoked only once during the encoding process.
-func ToEvent(ctx context.Context, message MessageReader, transformers TransformerFactories) (e event.Event, err error) {
-	e = event.New()
-
+func ToEvent(ctx context.Context, message MessageReader, transformers TransformerFactories) (*event.Event, error) {
```
Comment (slinkydeveloper) [inline pkg/binding/to_event.go]: I like it

File: pkg/binding/buffering/copy_message.go [hunk 2]
```diff
@@ -32,7 +32,7 @@ func CopyMessage(ctx context.Context, m binding.Message, transformers binding.Tr
 		if err != nil {
 			return nil, err
 		}
-		return binding.EventMessage(e), nil
+		return binding.EventMessage(*e), nil
```
Comment (slinkydeveloper) [inline pkg/binding/buffering/copy_message.go]: This provokes a copy, you should do `(*binding.EventMessage)(e)` to just cast it as a pointer

File: pkg/binding/buffering/copy_message.go [hunk 3]
```diff
@@ -48,6 +48,6 @@ func CopyMessage(ctx context.Context, m binding.Message, transformers binding.Tr
...
```
Comment (slinkydeveloper) [inline pkg/binding/buffering/copy_message.go]: This provokes a copy, you should do `(*binding.EventMessage)(e)` to just cast it as a pointer

File: pkg/binding/example_implementing_test.go [hunk 4]
```diff
@@ -43,7 +44,7 @@ type ExSender struct {
 	transformers binding.TransformerFactories
 }
 
-func NewExSender(w io.Writer, factories ...binding.TransformerFactory) binding.Sender {
+func NewExSender(w io.Writer, factories ...binding.TransformerFactory) bindings.Sender {
```
Comment (slinkydeveloper) [inline pkg/binding/example_implementing_test.go]: I think we should get rid of those tests or refactor them somehow, they are meant to explain to bindings implementers how to create a new binding. But I think it's more self explicative the http implementation honestly, even if it's a bit more complex...

File: pkg/binding/write.go [hunk 5]
```diff
@@ -86,13 +86,13 @@ func Write(
...
 
-	var e event.Event
+	var e *event.Event
 	e, err = ToEvent(ctx, message, transformers)
 	if err != nil {
 		return enc, err
 	}
 
-	message = EventMessage(e)
+	message = EventMessage(*e)
```
Comment (slinkydeveloper) [inline pkg/binding/write.go]: This provokes a copy, you should do `(*EventMessage)(e)` to just cast it as a pointer

File: pkg/bindings/kafka_sarama/producer_message_encoder.go [hunk 6]
```diff
@@ -76,7 +76,7 @@ func WriteKafkaProducerMessage(ctx context.Context, m binding.Message, producerM
 		producerMessage.Key = sarama.StringEncoder(s)
 	}
 
-	eventMessage := binding.EventMessage(e)
+	eventMessage := binding.EventMessage(*e)
```
Comment (slinkydeveloper) [inline pkg/bindings/kafka_sarama/producer_message_encoder.go]: This provokes a copy, you should do `(*binding.EventMessage)(e)` to just cast it as a pointer

File: pkg/bindings/test/test_test.go [hunk 7]
```diff
@@ -7,13 +7,14 @@ import (
 	"github.com/stretchr/testify/assert"
 
 	"github.com/cloudevents/sdk-go/pkg/binding"
-	"github.com/cloudevents/sdk-go/pkg/binding/test"
+	. "github.com/cloudevents/sdk-go/pkg/binding/test"
+	. "github.com/cloudevents/sdk-go/pkg/bindings/test"
 )
 
 func TestEvent(t *testing.T) {
```
Comment (slinkydeveloper) [inline pkg/bindings/test/test_test.go]: IMO This should continue to live in `"github.com/cloudevents/sdk-go/pkg/binding/test"`

File: pkg/bindings/test/transformer.go [hunk 8]
```diff
@@ -7,6 +7,7 @@ import (
 	"github.com/stretchr/testify/require"
```
Comment (slinkydeveloper) [inline pkg/bindings/test/transformer.go]: IMO This file should continue to live in `"github.com/cloudevents/sdk-go/pkg/binding/test"` (`RunTransformerTestArgs` and `RunTransformerTests`)

File: pkg/bindings/transport.go [hunk 9]
```diff
@@ -105,19 +106,19 @@ func (t *BindingTransport) handle(ctx context.Context, m Message) (err error) {
...
+	if err := t.handler.Receive(ctx, *e, &eventResp); err != nil {
 		return err
 	}
 
 	if eventResp.Event != nil {
 		// TODO: this does not give control over the http response code at the moment.
-		if rs, ok := m.(ResponseMessage); ok {
-			rs.Response(ctx, EventMessage(*eventResp.Event))
+		if rs, ok := m.(binding.ResponseMessage); ok {
+			rs.Response(ctx, binding.EventMessage(*eventResp.Event))
```
Comment (slinkydeveloper) [inline pkg/bindings/transport.go]: copy cast

File: pkg/binding/example_implementing_test.go (hunk 4 above)
Comment (n3wscott) [reply pkg/binding/example_implementing_test.go]: In the next PR I moved them

File: pkg/bindings/test/test_test.go (hunk 7 above)
Comment (n3wscott) [reply pkg/bindings/test/test_test.go]: These move in the next PR, I do a split

**Pull request closed without merging**